File: 29. Divide Two Integers/Solution.py
class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        if dividend == 0:
            return 0
        if dividend == -2**31 and divisor == -1:
            return 2**31 - 1
        positivie = (dividend > 0 and divisor > 0) or (dividend < 0
                                                       and divisor < 0)
        a = abs(dividend)
        b = abs(divisor)
        ans = 0
        while a >= b:
            shift = 0
            while a >= (b << shift):
                shift += 1
            a -= b << (shift - 1)
            ans += 1 << (shift - 1)
        return ans if positivie else -ans

File: 29. Divide Two Integers/Solution.cpp
class Solution
{
  public:
    int divide(int dividend, int divisor)
    {
        if (dividend == 0)
            return 0;
        if (dividend == INT_MIN && divisor == -1)
            return INT_MAX;
        bool positive = (dividend > 0 && divisor > 0) || (dividend < 0 && divisor < 0);
        int ans = 0;
        long a = abs((long)dividend);
        long b = abs((long)divisor);
        while (a >= b)
        {
            int shift = 0;
            while (a >= (b << shift))
            {
                shift++;
            }
            a -= b << (shift - 1);
            ans += 1 << (shift - 1);
        }
        return positive ? ans : -ans;
    }
};